src/Data/List/Ext/Properties.agda

```diff
@@ -54,11 +249,324 @@ module _ {a} {A : Set a} ⦃ _ : DecEq A ⦄ where
+                               (concatMap⁺ {xs = l ∷ []} λ l'∈ → case l'∈ of λ where (here refl) → l∈)
+
+  insert-decomp∈⁻¹' : {ls : List (List A)} {y : A} → ∀ xs
+                      → ∃[ ll ] ∃[ lr ] (xs ≡ ll ++ [ y ] ++ lr × ll ++ lr ∈ ls)
+                      → xs ∈ (concatMap (insert y) ls)
+  insert-decomp∈⁻¹' {.(ll ++ lr) ∷ ls} {y} .(ll ++ [ y ] ++ lr) (ll , lr , refl , here refl) =
+    insert⊆concat (ll ++ lr) (here{xs = ls} refl) insert-decomp≡⁻¹
+
+  insert-decomp∈⁻¹' {_ ∷ ls} {y} .(ll ++ [ y ] ++ lr) (ll , lr , refl , there lllr∈) =
+    concatMap⁺{xs = ls} there (insert-decomp∈⁻¹' (ll ++ y ∷ lr) (ll , lr , refl , lllr∈))
+
+
+  insert-decomp∈⁻¹ : {ls : List (List A)} {y : A} → ∀ xs → ∃[ l ] (l ∈ ls × xs ∈ (insert y l))
+                         → xs ∈ (concatMap (insert y) ls)
+  insert-decomp∈⁻¹ {l ∷ _} xs (.l , here refl , xs∈) = xs⊆xs++ys _ _ xs∈
+  insert-decomp∈⁻¹ {_ ∷ _} xs (l' , there l'∈ , xs∈) = xs⊆ys++xs _ _ (insert-decomp∈⁻¹ xs (l' , l'∈ , xs∈))
+
+  insert++distr : {ll lr : List (List A)} {xs : List A} {x : A} → xs ∈ concatMap (insert x) (ll ++ lr)
+                  → xs ∈ concatMap (insert x) ll ++ concatMap (insert x) lr
+
+  insert++distr {[]} xs∈xlr = xs∈xlr
+  insert++distr {ll ∷ lls} {[]} {xs} {x} xs∈xlr = case (∈-++⁻ (insert x ll) xs∈xlr) of λ where
+    (inj₁ v) → xs⊆xs++ys _ _ (xs⊆xs++ys (insert x ll) _ v)
+    (inj₂ v) → xs⊆xs++ys _ _ (xs⊆ys++xs _ (insert x ll)
+                              (subst (λ u → xs ∈ concatMap (insert x) u) (++-identityʳ lls) v))
+  insert++distr {ll ∷ lls} {lr ∷ lrs} {xs} {x} xs∈xlr = case (∈-++⁻ (insert x ll) xs∈xlr) of λ where
+    (inj₁ v) → xs⊆xs++ys _ _ (xs⊆xs++ys (insert x ll) _  v)
+    (inj₂ v) → case (∈-++⁻ (concatMap(insert x) lls) (insert++distr{lls} v)) of λ where
+               (inj₁ w) → xs⊆xs++ys (insert x ll ++ concatMap (insert x) lls) _ (xs⊆ys++xs _ (insert x ll) w)
+               (inj₂ w) → xs⊆ys++xs _ (insert x ll ++ concatMap (insert x) lls) w
+
+  insert++ʳ : {ll lr : List (List A)} {xs : List A} {x : A}
+              → xs ∈ concatMap (insert x) lr → xs ∈ concatMap (insert x) (ll ++ lr)
+  insert++ʳ {[]} xs∈xlr = xs∈xlr
+  insert++ʳ {ll ∷ lls} xs∈xlr = xs⊆ys++xs _ (insert _ ll) (insert++ʳ{ll = lls} xs∈xlr)
+
+  insert++distr⁻¹ : {ll lr : List (List A)} {x : A} → ∀ l
+                    → l ∈ concatMap (insert x) ll ++ concatMap (insert x) lr
+                    → l ∈ concatMap (insert x) (ll ++ lr)
+  insert++distr⁻¹ {[]} l l∈++ = l∈++
+  insert++distr⁻¹ {ll ∷ lls} {lr} l l∈++ with ∈-++⁻ (insert _ ll ++ concatMap (insert _) lls) l∈++
+  ...| inj₂ v = xs⊆ys++xs _ _ (insert++ʳ{ll = lls} v)
+  ...| inj₁ v =
+    case (∈-++⁻ _ (subst (λ u → l ∈ u)
+                    (++-assoc (insert _ ll) (concatMap (insert _) lls) (concatMap (insert _) lr))
+                    (xs⊆xs++ys _ _ v))) of λ where
+      (inj₁ w) → xs⊆xs++ys (insert _ ll) _ w
+      (inj₂ w) → xs⊆ys++xs _ _ (insert++distr⁻¹{ll = lls} l w)
```

Comment:
All of this is completely unnecessary, just use `concatMap-++`

Reply:
(replaced with `concatMap-++`)